src/page/lru.rs

```diff
@@ -18,139 +15,130 @@ impl Lru {
+    accesses: u64,
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone)]
 struct Shard {
-    accesses: BTreeMap<i64, Entry>,
-    entries: HashMap<LogID, Entry>,
+    accesses: BTreeMap<Duration, PageID>,
+    entries: HashMap<PageID, Entry>,
+    shadow: HashMap<PageID, Entry>,
+    capacity: usize,
+    sz: usize,
+    ctime: Instant,
 }
 
 impl Shard {
-    fn len(&self) -> usize {
-        self.entries.len()
+    fn new(capacity: usize) -> Shard {
+        Shard {
+            accesses: BTreeMap::new(),
+            entries: HashMap::new(),
+            shadow: HashMap::new(),
+            capacity: capacity,
+            sz: 0,
+            ctime: Instant::now(),
+        }
     }
 
-    fn insert(&mut self, mut entry: Entry) {
-        loop {
-            if self.accesses.contains_key(&entry.mtime) {
-                entry.mtime += 1;
-                continue;
-            }
-
-            self.accesses.insert(entry.mtime, entry.clone());
-            break;
+    fn pop(&mut self, pid: PageID) -> Option<Entry> {
+        let shadow = self.shadow.remove(&pid);
+        if shadow.is_some() {
+            return shadow;
         }
-        self.entries.insert(entry.lid, entry);
-    }
 
-    fn bump(&mut self, lid: LogID) {
-        let mtime = time::get_time().sec;
-        let entry_opt = self.entries.remove(&lid);
-        if entry_opt.is_none() {
-            // TODO we can't unwrap, but we should reduce this incidence
-            return;
+        if let Some(entry) = self.entries.remove(&pid) {
+            self.accesses.remove(&entry.mtime);
+            return Some(entry);
         }
-        let mut entry = entry_opt.unwrap();
-        entry.accesses.fetch_add(1, Ordering::SeqCst);
 
-        let old_mtime = entry.mtime;
-        entry.mtime = mtime;
+        None
+    }
+
+    fn insert(&mut self, entry: Entry) {
+        self.accesses.insert(entry.mtime, entry.pid);
+        self.entries.insert(entry.pid, entry);
+    }
 
-        self.accesses.remove(&old_mtime).unwrap();
-        loop {
-            if self.accesses.contains_key(&entry.mtime) {
-                entry.mtime += 1;
-                continue;
+    fn accessed(&mut self, pid: PageID, sz: usize) -> Vec<PageID> {
+        let mut entry = self.pop(pid).unwrap_or_else(|| {
+            Entry {
+                pid: pid,
+                sz: 0,
+                accesses: 0,
+                mtime: self.ctime.elapsed(),
             }
+        });
 
-            self.accesses.insert(entry.mtime, entry.clone());
-            break;
-        }
-        self.entries.insert(entry.lid, entry);
-    }
+        self.sz -= entry.sz;
 
-    fn pop(&mut self, lid: LogID) -> Result<(), ()> {
-        let entry_opt = self.entries.remove(&lid);
-        if entry_opt.is_none() {
-            return Err(());
-        }
+        entry.accesses += 1;
+        entry.sz = sz;
+        // guaranteed to be higher than any other time measured before
+        entry.mtime = self.ctime.elapsed();
```

Comment:
by calling `elapsed` here, we are guaranteed to get a higher number than has ever been returned before for this `Shard`. This is because the result is guaranteed to be monotonic. So, we don't need to worry about overwriting a different cache entry when we insert the entry into the `self.accesses` `BTreeMap`